### ISSUE TITLE
Heatmaps or Attention maps using GRAD-CAM with YOLOv8

### ISSUE THREAD

**Issue by subhanmangi** [ultralytics/ultralytics#2020]
Can anybody help how to use GRAD-CAM or CAM with YOLOv8 for creating heatmaps or attention maps?

**Comment by rigvedrs** [comment]
@subhanmangi you can use my implementation here https://github.com/rigvedrs/YOLO-V8-CAM . You can easily start using it for the yolo v8 object detection and classification models.

**Comment by rigvedrs** [comment]
@glenn-jocher I tried to search for the gradcam.py module you mentioned but could not find it. Can you provide a link to it? Thank you for your time.

**Comment by msadoun** [comment]
Did anyone find gradcam.py ???

**Comment by glenn-jocher** [comment]
@msadoun hi there! Thank you for reaching out. In the Ultralytics YOLOv8 repository, there isn't a specific `gradcam.py` module included. However, you can implement the Grad-CAM approach using the PyTorch ecosystem.

Grad-CAM involves computing the gradients of the target class output with respect to the network's feature maps. These gradients are then used to weigh the feature maps, creating an attention map. This attention map highlights the regions of the image that were influential in the final classification decision.

Regarding the implementation, you can refer to the PyTorch documentation or external resources to understand the details of Grad-CAM and how to integrate it with YOLOv8. There are various open-source implementations available, which you can adapt for your specific use case.

Please feel free to ask if you have any further questions or need additional assistance. We're here to help!

**Comment by Dammy-Joe** [comment]
Thanks @glenn-jocher for the explanation. I have found the GradCAM PyTorch implementation in the link you provided, https://github.com/1Konny/gradcam_plus_plus-pytorch.
The implementation is intended majorly for classification-based architectures (VGG, AlexNet, DenseNet etc.), to tweak it for YoLoV8, one would need to add details of the YoLoV8 layers to the utils.py source code, I suppose. Having checked through the YOLOv8 docs on ultralytics.com, I'm yet to find a way to do this. Is there a link you could point me to to get this done?

Also, I noticed that IoU (Intersection Over Union) is not part of the metrics produced as output after training YOLOv8 for detection. Although I'm hoping that at the testing phase, predicted bounding boxes will be saved to file and one can calculate IoU from that. Please, clarify this as well. Thanks.

**Comment by glenn-jocher** [comment]
@Dammy-Joe hi there, thanks for reaching out!

Regarding your question about Grad-CAM, you're correct. You would need to include additional lines of code that account for the specifics of the YOLOv8 architecture in the Grad-CAM PyTorch implementation. Unfortunately, we don't have a direct guide for this process but the main idea revolves around grabbing the feature maps from the model layers you're interested in and then applying the Grad-CAM approach to them.

Regarding your question about IoU, your understanding is correct as well. It isn't directly calculated and logged during training, although it plays a crucial role in the loss computation and therefore model adjustments. However, you're correct that once you have the bounding box predictions from testing phase, you can calculate IoU post-training, to further investigate and quantify the model's performance.

I hope this helps! Let us know if you have more questions.

**Comment by Dammy-Joe** [comment]
Thanks @glenn-jocher for your timely response. I've been playing around with the gradcam.py code and I'm trying to adapt it to YOLOv8. But I'll need to know the names of the layers in the YOLOv8 architecture. I just want to know if it's safe to assume that those names displayed in the architecture during training can be used as the names of the layers in YOLOv8 (see the attached screenshot).

Also, I found a discussion here: https://community.ultralytics.com/t/visualize-feature-maps/170/3 on visualizing feature maps for YOLOv5. If this is easier, I could use it as I'm only interested in seeing the feature maps of a few layers in the detection process and I can still make do with maps coming from YOLOv5, while still using YOLOv8 for the actual detection. So, would you say there's more support for feature map visualisation in YOLOv5 than in YOLOv8?

Thanks for your efforts.

<img width="549" alt="Screenshot 2023-09-26 124654" src="https://github.com/ultralytics/ultralytics/assets/60466997/73d6e0ba-b7ea-44b0-999f-ce953d43649a">

**Comment by glenn-jocher** [comment]
@Dammy-Joe, 

When observing the output of your model's summary during training, the names of the layers listed there can indeed be used as references to the layers of the YOLOv8 architecture.

As for visualizing the feature maps, YOLOv8 supports visualization as much as other versions. It's not necessarily easier or harder in different versions; the process would involve essentially the same steps. You would need to extract the feature maps from whichever layers you are interested in, and then plot them accordingly. 

Let us know if there's anything else you need help with!

**Comment by Cheryl33990** [comment]
@glenn-jocher I found a github combine yolov8 with grad-cam successfully. https://github.com/z1069614715/objectdetection_script/blob/master/yolo-gradcam/yolov8_heatmap.py

I just change the get_params function to my trained yolov8 detection model and the yaml file of my training data.
```python
def get_params():
    params = {
        'weight': 'model3_2cls.pt',  # change here
        'cfg':  'model3_2cls.yaml',  # change here
        'device': 'cuda:0',
        'method': 'GradCAM', # GradCAMPlusPlus, GradCAM, XGradCAM
        'layer': 'model.model[7]',
        'backward_type': 'all', # class, box, all
        'conf_threshold': 0.3, 
        'ratio': 0.02 # 0.02-0.1 
    }
    return params

if __name__ == '__main__':
    model = yolov8_heatmap(**get_params())
    model(r'small_0_0_8b134278-IMG_2195_JPG.rf.532f1f9c1d89fdf3723b06f4a581971c.JPG', 'result')
```
However, the error massage comes:
```python
                   from  n    params  module                                       arguments                     
Traceback (most recent call last):
  File "c:\01_project_yolo\yolov8_CAM\yolov8_heatmap.py", line 158, in <module>
    model = yolov8_heatmap(**get_params())
  File "c:\01_project_yolo\yolov8_CAM\yolov8_heatmap.py", line 57, in __init__
    model = Model(cfg, ch=3, nc=len(model_names)).to(device)
  File "c:\01_project_yolo\yolov8_CAM\ultralytics\nn\tasks.py", line 232, in __init__
    self.model, self.save = parse_model(deepcopy(self.yaml), ch=ch, verbose=verbose)  # model, savelist
  File "c:\01_project_yolo\yolov8_CAM\ultralytics\nn\tasks.py", line 677, in parse_model
    for i, (f, n, m, args) in enumerate(d['backbone'] + d['head']):  # from, number, module, args
KeyError: 'backbone'
```

I don't know why the error comes😢

**Comment by glenn-jocher** [comment]
@Cheryl33990 hi there!

From the error message, it looks like the code is unable to find the 'backbone' key in the dictionary returned by `parse_model()`. This suggests that your 'model3_2cls.yaml' file might be structured differently than what the function expects.

Make sure your .yaml file has sections labeled 'backbone' and 'head'. The 'backbone' section should contain the definitions of your base network, and the 'head' section should define the additional layers specific to the object detection task.

If you structured your 'model3_2cls.yaml' differently or used different section names, you would need to update the function to correctly parse your custom .yaml file.

This should solve the KeyError you're seeing. Let's know if you have any other questions!

**Comment by Cheryl33990** [comment]
@glenn-jocher Thanks for your response! 
I think I misunderstood what cfg means. 
(I originally thought that was referring to the YAML file for dataset.)

---
I have a new question below:
In my process, I used yolov8m.pt to train my customed model and get model3_2cls.pt.
(I didn't change the structure, just use it.)

Then, I checked the arg.yaml of train:
![image](https://github.com/ultralytics/ultralytics/assets/87057429/fd0fe9b7-0676-4001-966e-cdeda673c5c8)
I suppose that if I have the configuration file (cfg) of my model, I could try to add Grad-CAM.
 **How could I get model3_2cls.yaml (cfg)?**

**Comment by Cheryl33990** [comment]
@glenn-jocher I finish the problem!
Just change the nc in yolov8.yaml, and change the parameters in the function get_params.
Thanks for your help!

**Comment by glenn-jocher** [comment]
@Cheryl33990, that's great to hear! I'm glad you were able blend Grad-CAM with YOLOv8 successfully by adjusting the number of classes (nc) in yolov8.yaml and modifying the parameters in the get_params function. Keep exploring, and don't hesitate to ask if you have any more questions or run into any difficulties in the future. Happy coding!

**Comment by AndywithCV** [comment]
> @glenn-jocher I finish the problem! Just change the nc in yolov8.yaml, and change the parameters in the function get_params. Thanks for your help!

Hello, Can you kindly share the code for integrading Grad-Cam to YOLOv8, i am really thankful for it!!

**Comment by glenn-jocher** [comment]
@AndywithCV hello, 

Unfortunately, we can't directly share the diversified code, but I can surely guide you throughout the process. 

To integrate Grad-Cam with YOLOv8, first, you need to modify YOLOv8’s architecture file (yolo_v8.yaml). Particularly, in this file, you need to change the number of classes according to your dataset (denoted by 'nc'). 

Now, for integrating Grad-CAM with the model, you already have a generic Grad-CAM implementation. What you need to do is to adapt it to YOLOv8. Update your 'get_params' method to load your trained YOLOv8 model which would be using Grad-CAM for visualization. 

The Grad-CAM implementation should therefore recognize your custom YOLO model, and then plot feature maps. Hope this information would be helpful to you! Let me know if you face any issues.

**Comment by dimpolitik** [comment]
Hi all!

Should we care what value shoud we add in the layer attributre? I see `'layer': 'model.model[7]',` . However, in the source I see `'layer': 'model.model[9]',` 

In my case, I use YOLOv8 aftet a new training and the class yolov8_heatmap runs (i.e., no errors), but I don't get any result.

**Comment by dimpolitik** [comment]
Hi @Cheryl33990! 
I managed to implement the "class yolov8_heatmap" for my case study  using the link:https://github.com/z1069614715/objectdetection_script/blob/master/yolo-gradcam/yolov8_heatmap.py

Do you an idea why do we get so many images in the results directory? Which ones should we finally use for our analysis?
There is a for loop...

` for i in trange(int(post_result.size(0) * self.ratio)):
...
cam_image.save(f'{save_path}/{i}.png')
`

**Comment by glenn-jocher** [comment]
@dimpolitik hi there!

When using GradCAM with YOLOv8, the large number of images in the results directory is due to multiple heatmap overlays being created for different regions of the original image.

The 'for' loop you're referring to is iterating over a portion of the output tensor (determined by `int(post_result.size(0) * self.ratio)`), and for each iteration, it is creating and saving a heatmap overlay image. 

Which images to use for your analysis depends on your particular task. If you wish to analyze regions of the image where the model focuses most, you might want to consider the images corresponding to higher values in the output tensor. On the other hand, if you wish to get an overall understanding of where the model is focusing throughout the entire image, you might want to look at all the generated heatmaps.

The best approach is to explore different heatmaps and interpret them in the context of your specific use case.

**Comment by dimpolitik** [comment]
@glenn-jocher, very thoughtful explanation! Thank you!

**Comment by glenn-jocher** [comment]
@dimpolitik you're welcome! I'm glad to hear you found the explanation helpful. If you have further questions or run into any more issues, feel free to ask. I'm here to help. Happy coding!

**Comment by Rusab** [comment]
Hi @glenn-jocher, it is possible to generate heatmaps for different layers. Which layer should I generate the heatmap for if I want to explain my final results?

**Comment by glenn-jocher** [comment]
@Rusab hi there! For explaining your final results with heatmaps, you typically choose the layers closer to the output, as they contain higher-level features that contribute directly to the model's decision. However, examining heatmaps from various layers can give a more comprehensive view of what the model is learning at different levels of abstraction. Keep experimenting with different layers to see which provides the most interpretability for your use case. Happy analyzing!

**Comment by Mehreganbeiki** [comment]
Thank you for providing your GradCAM code. I am interested in utilizing your code to generate heatmaps for specific classes within my dataset. Could you please guide me on how to modify your code to achieve this?

**Comment by fatemehmomeni80** [comment]
Hi,
I want to implement GradCAM on YOLOv8 using this code:
https://github.com/z1069614715/objectdetection_script/blob/master/yolo-gradcam/yolov8_heatmap.py

I only changed this part: 
`def get_params():
    params = {
        'weight': '../heatmap_yolov8/yolov8s.pt', 
        'device': 'cpu',
        'method': 'GradCAM', 
        'layer': [10, 12, 14, 16, 18],
        'backward_type': 'all', 
        'conf_threshold': 0.2, # 0.2
        'ratio': 0.02, # 0.02-0.1
        'show_result': True, # 不需要绘制结果请设置为False
        'renormalize': False, # 需要把热力图限制在框内请设置为True(仅对detect,segment,pose有效)
        'task':'detect', # 任务(detect,segment,pose,obb,classify)
        'img_size':640, # 图像尺寸
    }
    return params

# pip install grad-cam==1.5.4 --no-deps
if __name__ == '__main__':
    model = yolo_heatmap(**get_params())
    model(r'./images.jpg', 'result')
`
Now when I try to run it, this image: 

![Image](https://github.com/user-attachments/assets/5870c03b-b910-4cad-8eb1-768d74a4a645)

shows output like this:

![Image](https://github.com/user-attachments/assets/ea33aab4-219f-48cc-aeed-d0f77cdaab7c)

What should I do???

Thank you..

**Comment by glenn-jocher** [comment]
Hi there! For class-specific heatmaps with YOLOv8, you can use our built-in `Heatmap` solution by specifying the `classes` parameter. Here's a minimal example:

```python
from ultralytics import solutions

heatmap = solutions.Heatmap(model="yolov8s.pt", classes=[0])  # class 0 only
results = heatmap(your_image)
```

This will generate heatmaps specifically for your chosen classes. For full implementation details, see the [Ultralytics Heatmap Documentation](https://docs.ultralytics.com/guides/heatmaps/). 

The output you're seeing appears normal - the heatmap intensity corresponds to object concentration areas. You can adjust the `colormap` parameter or intensity thresholds if you need different visualization characteristics.

**Comment by fatemehmomeni80** [comment]
Thank you for your response
No, I don't just want a heatmap
I'm studying the XAI field
I want to implement Gradcam and Gradcam++ and... on Yolov8
Thank you for your help

‫‪Glenn Jocher‬‏ ***@***.***‬‏> در تاریخ جمعه ۲۸ مارس ۲۰۲۵
ساعت ۲۱:۲۲ نوشت:‬

**Comment by glenn-jocher** [comment]
Hi! For Grad-CAM and other XAI methods with YOLOv8, you'll need to access intermediate layer activations and gradients. While we don't have built-in Grad-CAM support, you can implement it by:  

1. Using `model.model` to access the backbone/neck layers  
2. Registering forward/backward hooks on target layers  
3. Combining activations and gradients per the Grad-CAM paper  

Our [Model API documentation](https://docs.ultralytics.com/reference/engine/model/) shows how to access layer outputs. For XAI implementation guidance, see our [Explainable AI Guide](https://docs.ultralytics.com/). Let us know if you hit specific implementation roadblocks! 🚀

**Comment by hariouat** [comment]
Hello, Please how to visualize heatmap for oriented bounding boxes.

**Comment by rejens** [comment]
@fatemehmomeni80  were you able to implement gradcam with yolo?

**Comment by trung-kien-pham** [comment]
Hi @glenn-jocher, I had this problem when i tried to run the YOLO heatmap.

```py 
import cv2
from ultralytics import solutions

heatmap = solutions.Heatmap(model="models/carton_model/weights/best.pt", colormap=cv2.COLORMAP_PARULA, classes=[0, 1])
results = heatmap('Raw_images/photo_0202_jpg.rf.ebc4a96687c27c72beb4d7233325990c.png')
```

Ultralytics Solutions: {'source': None, 'model': 'models/carton_model/weights/best.pt', 'classes': [0, 1], 'show_conf': True, 'show_labels': True, 'region': None, 'colormap': 12, 'show_in': True, 'show_out': True, 'up_angle': 145.0, 'down_angle': 90, 'kpts': [6, 8, 10], 'analytics_type': 'line', 'figsize': (12.8, 7.2), 'blur_ratio': 0.5, 'vision_point': (20, 20), 'crop_dir': 'cropped-detections', 'json_file': None, 'line_width': 2, 'records': 5, 'fps': 30.0, 'max_hist': 5, 'meter_per_pixel': 0.05, 'max_speed': 120, 'show': True, 'iou': 0.7, 'conf': 0.25, 'device': None, 'max_det': 300, 'half': False, 'tracker': 'botsort.yaml', 'verbose': True, 'data': 'images'}
Traceback (most recent call last):
File "d:\Vscode project\YOLO-V12-CAM\test.py", line 5, in
results = heatmap('Raw_images/photo_0202_jpg.rf.ebc4a96687c27c72beb4d7233325990c.png')
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ultralytics\utils\plotting.py", line 205, in init
if im.shape[2] == 1: # handle grayscale
^^^^^^^^
AttributeError: 'str' object has no attribute 'shape

**Comment by glenn-jocher** [comment]
You’re getting this error because `Heatmap` expects an image array, not a file path string. Load the image with OpenCV first and pass the `np.ndarray`:

```python
import cv2
from ultralytics import solutions

# Load image as array
im = cv2.imread("Raw_images/photo_0202_jpg.rf.ebc4a96687c27c72beb4d7233325990c.png")
assert im is not None, "Failed to read image"

heatmap = solutions.Heatmap(
    model="models/carton_model/weights/best.pt",
    colormap=cv2.COLORMAP_PARULA,
    classes=[0, 1],
)

results = heatmap(im)  # not the path string
cv2.imwrite("heatmap_out.png", results.plot_im)
```

For more Python usage examples and options (e.g. video streams, regions, colormaps), check the heatmap guide at [Heatmaps using Ultralytics YOLO](https://docs.ultralytics.com/guides/heatmaps/).

**Comment by enescicek** [comment]
Hi @glenn-jocher 

Are the answers given here really written by you or a real person, or are they being generated by artificial intelligence in the background?

**Comment by glenn-jocher** [comment]
@enescicek Some replies are written directly by maintainers and some are AI-assisted drafts used to speed up support (and then edited as needed); if anything looks unclear or incorrect, please reply and we’ll tighten it up.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25548](https://github.com/ultralytics/ultralytics/pull/25548), titled **“Replace predict feature-map dumps with class activation heatmaps.”**

Key changes include:

- Replaced raw layer-by-layer feature-map dumps with practical PyTorch class activation heatmaps using LayerCAM.
- Updated `visualize=True` during prediction to generate heatmap overlays alongside prediction outputs.
- Applied confidence and class filters, with support for visualizing a configurable number of top predictions.
- Limited activation-map generation to supported PyTorch workflows, excluding tensor sources, augmentation, and embedding workflows.
- Updated inference and documentation to reflect the new heatmap behavior.

These changes should provide more interpretable visual explanations by highlighting image regions that influence predictions, while ensuring visualizations focus on relevant detections.

The changes were implemented by [@Y-T-G](https://github.com/Y-T-G), with contributions from [@glenn-jocher](https://github.com/glenn-jocher).

To test the latest changes, install directly from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release.

Please test this update and share whether it resolves the issue. Thank you 🙏 for reporting it, and any further feedback is welcome if the problem persists.

**Comment by glenn-jocher** [comment]
@subhanmangi, the improvement from @Y-T-G in #25548 is now merged. `model.predict(..., visualize=True)` generates one class-specific LayerCAM overlay per image, using the prediction score gradients instead of the old layer-by-layer feature dumps.
